tests/storage/linstor: Added `test_linstor_sr_expand_host` and `test_linstor_sr_expand_disk` to test expand Linstor SR

rushikeshjadhav opened the pull request.

rushikeshjadhav committed:
tests/storage/linstor: Added `--expansion-sr-disk` pytest option for specifying or auto-detecting Linstor SR expansion disks.

Signed-off-by: Rushikesh Jadhav <rushikesh7@gmail.com>

rushikeshjadhav committed:
Enhanced LVM cleanup with `pvremove` on all volume group devices.

Signed-off-by: Rushikesh Jadhav <rushikesh7@gmail.com>

rushikeshjadhav committed:
Added `expand_lvm_on_hosts` to expand LVM of Linstor SR across hosts.
Added `test_linstor_sr_expand_disk` to detect & assert change in SR size.

The test:
- Ensures `--expansion-sr-disk` disk(s) are available on the hosts.
- Configures LVM onto it and integrates into Linstor SR pool.
- Expands SR capacity and validates the increase.

Signed-off-by: Rushikesh Jadhav <rushikesh7@gmail.com>

rushikeshjadhav committed:
Added `test_linstor_sr_expand_host` to validate the expansion of Linstor SR by integrating a new host into the storage pool.

The test:
- Ensures the new host has available disks and installs required Linstor packages.
- Configures LVM and integrates it into Linstor SR pool.
- Expands SR capacity and validates the increase.
- Handles cleanup by removing the node.

Signed-off-by: Rushikesh Jadhav <rushikesh7@gmail.com>

ydirson commented on tests/storage/linstor/test_linstor_sr.py:
```diff
@@ -2,7 +2,7 @@
 import time
 
-from .conftest import LINSTOR_PACKAGE
+from .conftest import GROUP_NAME, LINSTOR_PACKAGE, LINSTOR_RELEASE_PACKAGE
```
Those parameter additions should also appear in `data.py-dist`

ydirson commented on tests/storage/linstor/test_linstor_sr.py:
```diff
@@ -86,6 +86,150 @@ def test_snapshot(self, vm_on_linstor_sr):
+        disks = []
+        for h in sr.pool.hosts:
+            logging.info("Checking on {}".format(h.hostname_or_ip))
+            available_disks = h.available_disks()
```
ah ok, forget it then

rushikeshjadhav committed:
Added `raw_disk_is_available()` using `blkid` for reliable raw disk checks.
Added docstring explaining what this new method does vs `disk_is_available`.
Added type hints.

Signed-off-by: Rushikesh Jadhav <rushikesh7@gmail.com>

rushikeshjadhav commented on lib/host.py:
```diff
@@ -516,6 +516,9 @@ def disks(self):
         disks.sort()
         return disks
 
+    def raw_disk_is_available(self, disk):
+        return self.ssh_with_result(['blkid', '/dev/' + disk]).returncode == 2
+
```
Done

rushikeshjadhav commented on tests/storage/linstor/test_linstor_sr.py:
```diff
@@ -86,6 +86,150 @@ def test_snapshot(self, vm_on_linstor_sr):
+            available_disks = h.available_disks()
+            # We need the disk to be "raw" (non LVM_member etc) to use
+            available_disks = [disk for disk in available_disks if h.raw_disk_is_available(disk)]
+            exapnsion_sr_disk = pytestconfig.getoption("expansion_sr_disk")
```
Corrected the type.

rushikeshjadhav commented on tests/storage/linstor/test_linstor_sr.py:
```diff
@@ -86,6 +86,150 @@ def test_snapshot(self, vm_on_linstor_sr):
+                    raise e
+
+        # Need to ensure that linstor is healthy/up-to-date before moving ahead.
+        time.sleep(30) # Wait time for Linstor node communications to restore.
```
This is in discussion, will prefer to update it with a note as of now and then introduce a reliable mechanism.

rushikeshjadhav commented on tests/storage/linstor/test_linstor_sr.py:
```diff
@@ -86,6 +86,150 @@ def test_snapshot(self, vm_on_linstor_sr):
+        assert int(new_sr_size) > int(sr_size) and resized is True, \
+            f"Expected SR size to increase but got old size: {sr_size}, new size: {new_sr_size}"
+        logging.info("SR expansion completed")
+        vm.shutdown(verify=True)
```
ok, will add it in the test docstring.

rushikeshjadhav committed:
Added fixtures to use in `test_linstor_sr_expand_disk` and `test_linstor_sr_expand_host`

Signed-off-by: Rushikesh Jadhav <rushikesh7@gmail.com>

rushikeshjadhav commented on tests/storage/linstor/test_linstor_sr.py:
```diff
@@ -86,6 +86,150 @@ def test_snapshot(self, vm_on_linstor_sr):
+
+    @pytest.mark.small_vm
+    def test_linstor_sr_expand_host(self, linstor_sr, host, hostB1, provisioning_type,
+                                    storage_pool_name, vm_on_linstor_sr):
```
I have updated the `test_linstor_sr_expand_host` as well as `test_linstor_sr_expand_disk` to use fixtures.

rushikeshjadhav commented on tests/storage/linstor/test_linstor_sr.py:
```diff
@@ -86,6 +86,150 @@ def test_snapshot(self, vm_on_linstor_sr):
+                logging.info("Found Disk {}".format(disk))
+                device = '/dev/' + disk
+                try:
+                    h.ssh(['pvcreate', '-ff', '-y', device])
```
After trial, it seems without `-ff` the PV remains initialised but orphan, as in there is no VG on it. This is an issue to ensure that the device has become `raw`.
